cpp/src/arrow/compute/kernels/aggregate_test.cc

```diff
@@ -592,25 +652,51 @@ TEST(TestNullSumKernel, Basics) {
-  EXPECT_THAT(Sum(ChunkedArrayFromJSON(ty, {"[null]", "[]", "[null, null]"})),
-              ResultWith(null_result));
+  for (auto func : {Sum, SumChecked}) {
+    SCOPED_TRACE(func);
+    auto default_options = ScalarAggregateOptions::Defaults();
+    EXPECT_THAT(func(ScalarFromJSON(ty, "null"), default_options, nullptr),
+                ResultWith(null_result));
+    EXPECT_THAT(func(ArrayFromJSON(ty, "[]"), default_options, nullptr),
+                ResultWith(null_result));
+    EXPECT_THAT(func(ArrayFromJSON(ty, "[null]"), default_options, nullptr),
+                ResultWith(null_result));
+    EXPECT_THAT(func(ChunkedArrayFromJSON(ty, {"[null]", "[]", "[null, null]"}),
+                     default_options, nullptr),
+                ResultWith(null_result));
 
-  ScalarAggregateOptions options(/*skip_nulls=*/true, /*min_count=*/0);
-  EXPECT_THAT(Sum(ScalarFromJSON(ty, "null"), options), ResultWith(zero_result));
-  EXPECT_THAT(Sum(ArrayFromJSON(ty, "[]"), options), ResultWith(zero_result));
-  EXPECT_THAT(Sum(ArrayFromJSON(ty, "[null]"), options), ResultWith(zero_result));
-  EXPECT_THAT(Sum(ChunkedArrayFromJSON(ty, {"[null]", "[]", "[null, null]"}), options),
-              ResultWith(zero_result));
+    ScalarAggregateOptions options(/*skip_nulls=*/true, /*min_count=*/0);
+    EXPECT_THAT(func(ScalarFromJSON(ty, "null"), options, nullptr),
+                ResultWith(zero_result));
+    EXPECT_THAT(func(ArrayFromJSON(ty, "[]"), options, nullptr), ResultWith(zero_result));
+    EXPECT_THAT(func(ArrayFromJSON(ty, "[null]"), options, nullptr),
+                ResultWith(zero_result));
+    EXPECT_THAT(func(ChunkedArrayFromJSON(ty, {"[null]", "[]", "[null, null]"}), options,
+                     nullptr),
+                ResultWith(zero_result));
+
+    options = ScalarAggregateOptions(/*skip_nulls=*/false, /*min_count=*/0);
+    EXPECT_THAT(func(ScalarFromJSON(ty, "null"), options, nullptr),
+                ResultWith(null_result));
+    EXPECT_THAT(func(ArrayFromJSON(ty, "[]"), options, nullptr), ResultWith(zero_result));
+    EXPECT_THAT(func(ArrayFromJSON(ty, "[null]"), options, nullptr),
+                ResultWith(null_result));
+    EXPECT_THAT(func(ChunkedArrayFromJSON(ty, {"[null]", "[]", "[null, null]"}), options,
+                     nullptr),
+                ResultWith(null_result));
+  }
+}
 
-  options = ScalarAggregateOptions(/*skip_nulls=*/false, /*min_count=*/0);
-  EXPECT_THAT(Sum(ScalarFromJSON(ty, "null"), options), ResultWith(null_result));
-  EXPECT_THAT(Sum(ArrayFromJSON(ty, "[]"), options), ResultWith(zero_result));
-  EXPECT_THAT(Sum(ArrayFromJSON(ty, "[null]"), options), ResultWith(null_result));
-  EXPECT_THAT(Sum(ChunkedArrayFromJSON(ty, {"[null]", "[]", "[null, null]"}), options),
-              ResultWith(null_result));
+TEST(TestSumKernel, Overflow) {
+  int64_t large_scalar = 1000000000000000;
+  int64_t length = 10000;
+  auto scalar = std::make_shared<Int64Scalar>(large_scalar);
```

Comment:
Can we test overflow using more types (including decimals)? It should be relatively simple even without templating if you use `ScalarFromJSON`.

Also, this is testing overflow when summing a broadcast scalar, but we should also test overflow within an array.

Reply:
I added uint64, but I don't think there are checked arithmetic ops for decimals? [AddChecked](https://github.com/apache/arrow/blob/main/cpp/src/arrow/compute/kernels/base_arithmetic_internal.h#L85) and [MultiplyChecked](https://github.com/apache/arrow/blob/main/cpp/src/arrow/compute/kernels/base_arithmetic_internal.h#L356) do not seem to check for overflows.

Reply:
Also, the test is for the array generated by `ArrayFromScalar`, not the scalar itself. I tried to pass a scalar and a length but the length is overriden to 1 [here](https://github.com/apache/arrow/blob/main/cpp/src/arrow/compute/function.cc#L260). So I'm not sure how to test the broadcast scalar case...

Reply:
Ah, I misunderstood the test, sorry. If the broadcast scalar cannot be triggered by user code then I guess we can ignore it for now?

Reply:
> I added uint64, but I don't think there are checked arithmetic ops for decimals? [AddChecked](https://github.com/apache/arrow/blob/main/cpp/src/arrow/compute/kernels/base_arithmetic_internal.h#L85) and [MultiplyChecked](https://github.com/apache/arrow/blob/main/cpp/src/arrow/compute/kernels/base_arithmetic_internal.h#L356) do not seem to check for overflows.

That seems to be because Add and Multiply will adjust the result precision as required:
```python
>>> a = pa.array([Decimal(9000)])
>>> a.type
Decimal128Type(decimal128(4, 0))
>>> pc.add_checked(a, a)
<pyarrow.lib.Decimal128Array object at 0x7f399c4c78e0>
[
  18000
]
>>> pc.add_checked(a, a).type
Decimal128Type(decimal128(5, 0))
```

However, Sum does not:
```python
>>> a = pa.array([Decimal(1000)]*20)
>>> a.type
Decimal128Type(decimal128(4, 0))
>>> pc.sum(a)
<pyarrow.Decimal128Scalar: Decimal('20000')>
>>> pc.sum(a).validate(full=True)
Traceback (most recent call last):
  ...
ArrowInvalid: Decimal value 20000 does not fit in precision of decimal128(4, 0)
```

Reply:
I see. Let me see what can be done.

Reply:
Note that a simple way to address the issue would be to validate the result just I did in the snippet above. It would probably not require significant changes to the template machinery.

Reply:
That's true. But I think in the end we still need make 'sum' adjust the precision of the result. Should I go with validation approach for now and fix the precision later?

Reply:
In the case of `Add`, the precision is adjusted based purely on the input types, not values (so it's predictable).
In the case of `Sum`, we would have to vary the precision based on the values or at least the number of values, which doesn't seem very attractive. So I think it would be better to do overflow checking instead.

Reply:
I see. That makes sense!

Reply:
I added the validation code for decimal. I check the precision directly instead of calling `Validate`. This achieves the same thing but without the cost of type dispatching.